proof/invariant-abstract/DetSchedAux_AI.thy

```diff
@@ -130,106 +76,147 @@ lemma cap_insert_no_pred_tcb_at:
-  assumes invoke_untyped_etcb_at:
-    "\<And>P t ui.
-      \<lbrace>(\<lambda>s :: det_ext state. etcb_at P t s) and valid_etcbs\<rbrace>
-        invoke_untyped ui
-      \<lbrace>\<lambda>r s. st_tcb_at (Not o inactive) t s \<longrightarrow> etcb_at P t s\<rbrace> "
-  assumes init_arch_objects_valid_etcbs[wp]:
-    "\<And>t d r n sz refs. \<lbrace>valid_etcbs\<rbrace> init_arch_objects t d r n sz refs \<lbrace>\<lambda>_. valid_etcbs\<rbrace>"
+  assumes init_arch_objects_idle_thread[wp]:
+    "\<And>t d r n sz refs P. init_arch_objects t d r n sz refs \<lbrace>\<lambda>s::'state_ext state. P (idle_thread s)\<rbrace>"
+  assumes init_arch_objects_etcbs_of[wp]:
+    "\<And>t d r n sz refs P.
+      init_arch_objects t d r n sz refs \<lbrace>\<lambda>s::'state_ext state. P (etcbs_of s)\<rbrace>"
   assumes init_arch_objects_valid_blocked[wp]:
-    "\<And>t d r n sz refs. \<lbrace>valid_blocked\<rbrace> init_arch_objects t d r n sz refs \<lbrace>\<lambda>_. valid_blocked\<rbrace>"
-  assumes invoke_untyped_cur_domain[wp]:
-    "\<And>P i. \<lbrace>\<lambda>s. P (cur_domain s)\<rbrace> invoke_untyped i \<lbrace>\<lambda>_ s. P (cur_domain s)\<rbrace>"
-  assumes invoke_untyped_ready_queues[wp]:
-    "\<And>P i. \<lbrace>\<lambda>s. P (ready_queues s)\<rbrace> invoke_untyped i \<lbrace>\<lambda>_ s. P (ready_queues s)\<rbrace>"
-  assumes invoke_untyped_scheduler_action[wp]:
-    "\<And>P i. \<lbrace>\<lambda>s. P (scheduler_action s)\<rbrace> invoke_untyped i \<lbrace>\<lambda>_ s. P (scheduler_action s)\<rbrace>"
-
-lemma delete_objects_valid_etcbs[wp]: "\<lbrace>valid_etcbs\<rbrace> delete_objects a b \<lbrace>\<lambda>_. valid_etcbs\<rbrace>"
+    "\<And>t d r n sz refs. init_arch_objects t d r n sz refs \<lbrace>valid_blocked :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  assumes init_arch_objects_cur_domain[wp]:
+    "\<And>t d r n sz refs P. init_arch_objects t d r n sz refs \<lbrace>\<lambda>s::'state_ext state. P (cur_domain s)\<rbrace>"
+  assumes init_arch_objects_ready_queues[wp]:
+    "\<And>t d r n sz refs P. init_arch_objects t d r n sz refs \<lbrace>\<lambda>s::'state_ext state. P (ready_queues s)\<rbrace>"
+  assumes init_arch_objects_scheduler_action[wp]:
+    "\<And>t d r n sz refs P. init_arch_objects t d r n sz refs \<lbrace>\<lambda>s::'state_ext state. P (scheduler_action s)\<rbrace>"
+
+lemmas mapM_x_defsym = mapM_x_def[symmetric]
+
+lemma delete_objects_etcb_at[wp]:
+  "delete_objects a b \<lbrace>\<lambda>s. etcb_at P t s\<rbrace>"
   apply (simp add: delete_objects_def)
-  apply (wpsimp simp: detype_def detype_ext_def wrap_ext_det_ext_ext_def do_machine_op_def)
-  apply (simp add: valid_etcbs_def st_tcb_at_kh_def obj_at_kh_def obj_at_def is_etcb_at_def)
+  apply (wpsimp simp: detype_def)
+  apply (auto simp: detype_def etcbs_of'_def etcb_at'_def)
   done
 
-lemmas mapM_x_defsym = mapM_x_def[symmetric]
+crunch reset_untyped_cap
+  for etcb_at[wp]: "etcb_at P t"
```

Comment:
but surely `etcb_at` could now be just `tcb_at`, or am I missing something?

Reply:
Huh, yeah, that's definitely true. That does feel like a something we should do, as long as it doesn't take too long now to update everywhere. But then we would want to make the same change on `rt`, and I'm not sure if that's a priority right now.

Reply:
True. Worth noting in an issue, though, it's something we should remove at some point. Happy for that to be some other time.